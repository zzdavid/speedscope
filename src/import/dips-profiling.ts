import {CallTreeProfileBuilder, ProfileGroup, FrameInfo} from '../lib/profile'
import {ProfileDataSource, MultiFileDataSource} from '../import/utils'

// Custom import for DIPS profiling logs

interface ProfCall {
  // parsed
  context: string
  callID: string
  parentCallID: string
  rootCallID: string
  threadID: number
  at: number        // timestamp milliseconds
  elapsed: number   // milliseconds
  appID: string
  host: string
  // calculated
  parent: ProfCall | null
  children: ProfCall[]
}

interface ImportFile {
  calls: ProfCall[]
  callsTo: Map<ImportFile, number>
}

//const ColVersion = 0
const ColTimestamp = 1
//const ColEnvironment = 2
const ColAppID = 3
const ColHostName = 4
//const ColDotNetTaskId = 5
const ColThreadID = 6
const ColRootCallID = 7
const ColCallID = 8
const ColParentCallID = 9
//const ColCallLevel = 10
const ColElapsed = 11
//const ColReserved = 12
const ColContext = 13
//const ColSessionId = 14

class MyBuilder extends CallTreeProfileBuilder {
  lastEventAt : number = 0

  enterFrame(frameInfo: FrameInfo, value: number) {
    this.lastEventAt = value;
    super.enterFrame(frameInfo, value)
  }

  leaveFrame(frameInfo: FrameInfo, value: number) {
    this.lastEventAt = value;
    super.leaveFrame(frameInfo, value)
  }
}

function parseTimestamp(s : string): number {
  // Parse as local time and return number of milliseconds
  // since 1970-01-01
  return new Date(parseInt(s.substr(0, 4)),   // year
                  parseInt(s.substr(4, 2))-1, // month (0-11)
                  parseInt(s.substr(6, 2)),   // day
                  parseInt(s.substr(8, 2)),   // hours
                  parseInt(s.substr(10, 2)),  // minutes
                  parseInt(s.substr(12, 2)),  // seconds
                  parseInt(s.substr(14, 3))   // milliseconds
                  ).valueOf()
}

function parseCall(rawEvent: string): ProfCall {
  const cols = rawEvent.split(';')

  return {
    // parsed
    context: cols[ColContext],
    callID: cols[ColCallID],
    parentCallID: cols[ColParentCallID],
    rootCallID: cols[ColRootCallID],
    threadID: parseInt(cols[ColThreadID]),
    at: parseTimestamp(cols[ColTimestamp]),
    elapsed: parseInt(cols[ColElapsed]),
    appID: cols[ColAppID],
    host: cols[ColHostName],
    // calculated
    parent: null,
    children: []
  };
}

function setUpRelationsOnAllCalls(files: ImportFile[]) {
  let calls : ProfCall[] = []
  
  for (let file of files) {
    calls = calls.concat(file.calls)
  }

  setUpRelations(calls)
}

function setUpRelations(calls: ProfCall[]) {
  const callIdMap = new Map<string, ProfCall>()

  for (let call of calls) {
    callIdMap.set(call.callID, call)
  }

  for (let call of calls) {
    const parent = callIdMap.get(call.parentCallID)
    if (parent)
    {
      call.parent = parent
      parent.children.push(call)
    }
  }  
}

function timeOfFirstCall(calls: ProfCall[]) : number {
  // Find first timestamp. It should be the first top-level call
  // but we might as well check all
  let min = Number.MAX_SAFE_INTEGER

  for (let call of calls)
    if (call.at < min)
      min = call.at

  return min
}

function translateByOffset(files : ImportFile[], offset : number) {
  for (let file of files) {
    for (let call of file.calls) {
      call.at += offset
    }
  }
}

function getBuilderForCall(call: ProfCall, builders: MyBuilder[], width : number) : MyBuilder {
  // Profiles can't show simultaneous calls. Find a free profile
  // for this call or create a new one.
  let found = builders.find(bldr => bldr.lastEventAt <= call.at)

  if (!found) {
    found = new MyBuilder(width)
    found.setName(`Group ${builders.length + 1}`)
    builders.push(found)
  }

  return found
}

function adjustToInterval(value: number, min: number, max: number) : number {
  if (value < min) return min
  if (value > max) return max
  return value
}

function addFrames(call: ProfCall, builder: MyBuilder, parent_end: number) {
  const app = `App-${call.appID}@${call.host}`

  const frameInfo : FrameInfo = {
    key: app + ';' + call.context,
    name: call.context,
    file: app
  }

  // We need to adjust the times b/c timestamps and elapsed times are not
  // exact and sometimes there is a millisecond overlap, which would cause
  // the builder to fail
  let start = adjustToInterval(call.at, builder.lastEventAt, parent_end)
  let end = adjustToInterval(start + call.elapsed, start, parent_end)

  try {
    builder.enterFrame(frameInfo, start)

    for (let child of call.children) {
      addFrames(child, builder, end)
    }

    builder.leaveFrame(frameInfo, end)
  }
  catch (_e)
  {
    const err : Error = _e
    err.message += '\nError at CallID: ' + call.callID
    throw err
  }
}

async function parseFile(dataSource: ProfileDataSource): Promise<ImportFile> {
  let contents = await dataSource.readAsText()

  const calls = contents.split('\r\n')
                        .filter(s => s.startsWith('1;'))
                        .map(parseCall)

  return {
    calls: calls,
    callsTo: new Map<ImportFile, number>()
  }
}

function guessTopLevelFile(files: ImportFile[]): ImportFile {
  if (files.length == 1) {
    // quick return for performance
    return files[0]
  }

  let callMap = new Map()

  // Map callID -> file
  for (let file of files) {
    for (let call of file.calls) {
      callMap.set(call.callID, file)
    }
  }

  // Fill in callsTo
  for (let file of files) {
    for (let call of file.calls) {
      const parent = callMap.get(call.parentCallID)
      if (parent && parent !== file) {
        const n = parent.callsTo.get(file) || 0
        parent.callsTo.set(file, n+1)
      }
    }
  }

  // Find the first file in the call chain, i.e. from the service that calls other services
  let first = files[0]

  for (let file of files)
  {
    if ((file.callsTo.get(first) || 0) > (first.callsTo.get(file) || 0)) {
      first = file
    }
  }

  return first
}

function filterByRootIDs(files: ImportFile[], rootIDs : Set<string>) {
  for (var file of files)
    file.calls = file.calls.filter(c => rootIDs.has(c.rootCallID))
}

function toSet<T>(array : Array<T>) : Set<T> {
  const set = new Set<T>()
  for (var item of array) {
    set.add(item)
  }
  return set
}

export async function importDIPSProfiling(dataSource: ProfileDataSource): Promise<ProfileGroup | null> {
  let dataSources = (dataSource instanceof MultiFileDataSource) ?
                        (dataSource as MultiFileDataSource).dataSources : [dataSource]

  let files = await Promise.all(dataSources.map(d => parseFile(d)))
  
  files = files.filter(f => f.calls.length > 0)
  
  if (files.length == 0) {
    return null
  }

  // We'll only show calls from the first file in the call order. The 
  // others are called from this and are just adding frames to the profile.

  const primaryFile = guessTopLevelFile(files)
  const extraFiles = files.filter(f => f !== primaryFile)

  const rootCallIDs = toSet(primaryFile.calls.map(c => c.rootCallID))

  filterByRootIDs(extraFiles, rootCallIDs)

  const offset = -timeOfFirstCall(primaryFile.calls)

  translateByOffset(files, offset)

  setUpRelationsOnAllCalls(files)

  const topCalls = primaryFile.calls.filter(c => c.parent == null)
  const width = topCalls[topCalls.length-1].at + topCalls[topCalls.length-1].elapsed  
  const profiles : MyBuilder[] = []

  for (let call of topCalls) {
    const builder = getBuilderForCall(call, profiles, width)
    addFrames(call, builder, width)
  }

  return {
    name: '',
    indexToView: 0,
    profiles: profiles.map(p => p.build()),
  }
}

export function isDIPSProfiling(contents: string): boolean {
  if (!contents) {
    return false
  }

  if (!contents.startsWith('1;')) {
    return false
  }

  const row = contents.split('\r\n')[0]
  const cols = row.split(';')

  return cols.length == 21
}