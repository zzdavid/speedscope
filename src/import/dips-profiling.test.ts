import {checkProfileSnapshot} from '../lib/test-utils'
import { isDIPSProfiling } from './dips-profiling'

test('importDIPSProfiling simple', async () => {
  await checkProfileSnapshot('./sample/profiles/DIPS/simple-profiling.log')
})

test('importDIPSProfiling error out of order', async () => {
  // Failed with "Samples must be provided in increasing order of cumulative value. Last
  // sample was 10, this sample was 9" at U8uBT54VskG+Nrn4aduCOg
  await checkProfileSnapshot('./sample/profiles/DIPS/out-of-order-profiling.log')
})

test('isDIPSProfiling returns true on valid format', () => {
  const valid = 
`1;20191107120357753;NA;58;vt-irq0-srv01;;66;t4f1n7Ev/k6tIC1J8weZHQ;CsX7tNLaAE6LFxFnPYx1WA;zZztk7B6P0+bH37Uh826SA;3;47;;DFS/FS/AuditEventPublisher/PostAuditEventAsync.HttpCall;nG8uRYlMGEaPaMBJtaGnOA;;0;1573128237753346;1573128237801129;;
1;20191107120357751;NA;58;vt-irq0-srv01;;66;t4f1n7Ev/k6tIC1J8weZHQ;zZztk7B6P0+bH37Uh826SA;N3ps3DPsc0OAMe/iI8/0IA;2;49;;DFS/FS/AuditEventPublisher/PostAuditEventAsync;nG8uRYlMGEaPaMBJtaGnOA;;0;1573128237751684;1573128237801289;;`;
  expect(isDIPSProfiling(valid)).toBe(true);
})

test('isDIPSProfiling returns false on invalid format', () => {
  const invalid = "a;b;c;d;e;f;g;h;i;j;k;l;m;n;o;p;q;r;s;t;u;v;w;x;y;z 1";
  expect(isDIPSProfiling(invalid)).toBe(false);
})
