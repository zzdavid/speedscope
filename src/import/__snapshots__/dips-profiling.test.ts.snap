// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importDIPSProfiling 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditFhirFormatBatchPublisher/PublishAllAuditEventsAsync",
      "line": undefined,
      "name": "DFS/FS/AuditFhirFormatBatchPublisher/PublishAllAuditEventsAsync",
      "selfWeight": 0,
      "totalWeight": 49,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditFhirFormatBatchPublisher/PublishLogsInternalAsync",
      "line": undefined,
      "name": "DFS/FS/AuditFhirFormatBatchPublisher/PublishLogsInternalAsync",
      "selfWeight": 0,
      "totalWeight": 49,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditEventPublisher/PostAuditEventAsync",
      "line": undefined,
      "name": "DFS/FS/AuditEventPublisher/PostAuditEventAsync",
      "selfWeight": 2,
      "totalWeight": 49,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditEventPublisher/PostAuditEventAsync.HttpCall",
      "line": undefined,
      "name": "DFS/FS/AuditEventPublisher/PostAuditEventAsync.HttpCall",
      "selfWeight": 47,
      "totalWeight": 47,
    },
  ],
  "name": "Group 1",
  "stacks": Array [
    "DFS/FS/AuditFhirFormatBatchPublisher/PublishAllAuditEventsAsync;DFS/FS/AuditFhirFormatBatchPublisher/PublishLogsInternalAsync;DFS/FS/AuditEventPublisher/PostAuditEventAsync 2",
    "DFS/FS/AuditFhirFormatBatchPublisher/PublishAllAuditEventsAsync;DFS/FS/AuditFhirFormatBatchPublisher/PublishLogsInternalAsync;DFS/FS/AuditEventPublisher/PostAuditEventAsync;DFS/FS/AuditEventPublisher/PostAuditEventAsync.HttpCall 47",
  ],
}
`;

exports[`importDIPSProfiling error out of order 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;REST/Server//dips.oauth/identity/connect/token/POST",
      "line": undefined,
      "name": "REST/Server//dips.oauth/identity/connect/token/POST",
      "selfWeight": 45,
      "totalWeight": 56,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/ClientV3Store/FindClientByClientIdAsync",
      "line": undefined,
      "name": "DFS/FS/ClientV3Store/FindClientByClientIdAsync",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/ScopeV3Store/GetAllScopesAsync",
      "line": undefined,
      "name": "DFS/FS/ScopeV3Store/GetAllScopesAsync",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/ScopeV3Store/TryGetFromCacheAsync",
      "line": undefined,
      "name": "DFS/FS/ScopeV3Store/TryGetFromCacheAsync",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/ScopeV3Store/FindScopesAsync",
      "line": undefined,
      "name": "DFS/FS/ScopeV3Store/FindScopesAsync",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal",
      "line": undefined,
      "name": "DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal",
      "selfWeight": 2,
      "totalWeight": 9,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserFromIssuerSettingsAsync",
      "line": undefined,
      "name": "DFS/FS/IssuerIncomingClaimMappingResolver/GetUserFromIssuerSettingsAsync",
      "selfWeight": 0,
      "totalWeight": 2,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/IssuerIncomingClaimMappingResolver/GetClaimMappings",
      "line": undefined,
      "name": "DFS/FS/IssuerIncomingClaimMappingResolver/GetClaimMappings",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/IssuerIncomingClaimMappingResolver/GetClaimsFromPrincipal",
      "line": undefined,
      "name": "DFS/FS/IssuerIncomingClaimMappingResolver/GetClaimsFromPrincipal",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserByMappingClaim",
      "line": undefined,
      "name": "DFS/FS/IssuerIncomingClaimMappingResolver/GetUserByMappingClaim",
      "selfWeight": 0,
      "totalWeight": 2,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/UserDBStorage/GetUserBySignatureAsync",
      "line": undefined,
      "name": "DFS/FS/UserDBStorage/GetUserBySignatureAsync",
      "selfWeight": 1,
      "totalWeight": 3,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;Database",
      "line": undefined,
      "name": "Database",
      "selfWeight": 1,
      "totalWeight": 6,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;Database/ExecuteReader",
      "line": undefined,
      "name": "Database/ExecuteReader",
      "selfWeight": 4,
      "totalWeight": 4,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;Database/Read",
      "line": undefined,
      "name": "Database/Read",
      "selfWeight": 1,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserRoleForUser",
      "line": undefined,
      "name": "DFS/FS/IssuerIncomingClaimMappingResolver/GetUserRoleForUser",
      "selfWeight": 0,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/UserDBStorage/GetUserRolesBySignatureAsync",
      "line": undefined,
      "name": "DFS/FS/UserDBStorage/GetUserRolesBySignatureAsync",
      "selfWeight": 0,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/UserSessionCreator/GetOrCreateUserSession",
      "line": undefined,
      "name": "DFS/FS/UserSessionCreator/GetOrCreateUserSession",
      "selfWeight": 1,
      "totalWeight": 4,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditHelper/LogSessionCreationAsync",
      "line": undefined,
      "name": "DFS/FS/AuditHelper/LogSessionCreationAsync",
      "selfWeight": 1,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditHelper/GetDefaultIdentityServerAuditEvent",
      "line": undefined,
      "name": "DFS/FS/AuditHelper/GetDefaultIdentityServerAuditEvent",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditHelper/GetUserInfo",
      "line": undefined,
      "name": "DFS/FS/AuditHelper/GetUserInfo",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditServiceClientEventManager/CreateAuditEventAsync",
      "line": undefined,
      "name": "DFS/FS/AuditServiceClientEventManager/CreateAuditEventAsync",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditFhirFormatBatchPublisher/PublishAsync",
      "line": undefined,
      "name": "DFS/FS/AuditFhirFormatBatchPublisher/PublishAsync",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "App-58@vt-irq0-srv01",
      "key": "App-58@vt-irq0-srv01;DFS/FS/AuditFhirFormatBatchPublisher/PublishAsync.ToFhirAuditEvent",
      "line": undefined,
      "name": "DFS/FS/AuditFhirFormatBatchPublisher/PublishAsync.ToFhirAuditEvent",
      "selfWeight": 0,
      "totalWeight": 0,
    },
  ],
  "name": "Group 1",
  "stacks": Array [
    "REST/Server//dips.oauth/identity/connect/token/POST 8",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserFromIssuerSettingsAsync;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserByMappingClaim;DFS/FS/UserDBStorage/GetUserBySignatureAsync;Database;Database/ExecuteReader 1",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserFromIssuerSettingsAsync;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserByMappingClaim;DFS/FS/UserDBStorage/GetUserBySignatureAsync 1",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal 1",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal;DFS/FS/IssuerIncomingClaimMappingResolver/GetUserRoleForUser;DFS/FS/UserDBStorage/GetUserRolesBySignatureAsync;Database;Database/Read 1",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal;DFS/FS/UserSessionCreator/GetOrCreateUserSession 1",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal;DFS/FS/UserSessionCreator/GetOrCreateUserSession;Database;Database/ExecuteReader 3",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/IssuerIncomingClaimMappingResolver/GetSessionInternal 1",
    "REST/Server//dips.oauth/identity/connect/token/POST 1",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/AuditHelper/LogSessionCreationAsync 1",
    "REST/Server//dips.oauth/identity/connect/token/POST;DFS/FS/UserDBStorage/GetUserBySignatureAsync;Database 1",
    "REST/Server//dips.oauth/identity/connect/token/POST 36",
  ],
}
`;

exports[`importDIPSProfiling error out of order: indexToView 1`] = `0`;

exports[`importDIPSProfiling error out of order: profileGroup.name 1`] = `"out-of-order-profiling.log"`;

exports[`importDIPSProfiling: indexToView 1`] = `0`;

exports[`importDIPSProfiling: profileGroup.name 1`] = `"simple-profiling.log"`;
